JBL partybox-like soundreactive effect

tonyxforce opened the pull request:
this currently only works with the "color gradient" and the "colors 1&2" palettes

tonyxforce committed:
modified:   wled00/FX.cpp
	modified:   wled00/FX.h

tonyxforce committed:
modified:   wled00/FX.cpp

tonyxforce committed:
fixed slider names

arneboe commented on wled00/FX.cpp:
```diff
@@ -1915,6 +1915,96 @@ uint16_t mode_pride_2015(void) {
+//////////////////////
+
+
+uint16_t pos = 0;
+uint8_t hue = 0;
+int hueDelay = 0;
```
Maybe change the scope of those variables to be static inside `mode_jbl` or put them in a struct or change their names to something less generic.
It is very likely that other fx developers will use variables with the same name. In the best case scenario this will just spam a "shadows variable" warning. In the worst case they will use your variables by accident and strange stuff happens.

softhack007 replied:
see my proposal below ... I agree these should not be global. Better to use SEGENV.aux0, SEGENV.aux1, SEGENV.step.

softhack007 commented:
Hi, thanks for the PR :-)

Yes there is a lot of optimization potential with your code. In fact it should not depend on global variables, as effects may run in several instances in parallel, especially they will be used with several segments.

For a start:
* instead of global data, try to use "SEGENV.aux" variables.
* WLED has standard macros for extracting the R/G/B components from an uint32: R(x), G(x), B(x).

Don't worry, we will guide you through improving your code. 


https://github.com/MoonModules/WLED/blob/9e4f040d4103fd3069bcab633ad89d690f2bf1b4/wled00/FX.h#L393-L398

https://github.com/MoonModules/WLED/blob/9e4f040d4103fd3069bcab633ad89d690f2bf1b4/wled00/wled.h#L856-L861

Also this guide is worth reading (however a bit outdated): https://github.com/atuline/WLED/wiki/WLED-Programming-Notes#important-wled-variables


I'll make a few more proposals once I find a bit of time to read your source code in detail.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -1915,6 +1915,96 @@ uint16_t mode_pride_2015(void) {
+    hue++;
+  }
+
+  float speed = 0;
```
this should be an integer (instead of float), like `uint_fast32_t`

softhack007 commented on wled00/FX.cpp:
```diff
@@ -1915,6 +1915,96 @@ uint16_t mode_pride_2015(void) {
+  uint16_t counter = 0;
+
+  if (volumeSmth * 2 > (255 - SEGMENT.intensity))
+  {
```
coding style: opening braces should be on one line with the 'if'.

 https://github.com/MoonModules/WLED/blob/mdev/CONTRIBUTING.md#blocks

softhack007 commented on wled00/FX.cpp:
```diff
@@ -1915,6 +1915,96 @@ uint16_t mode_pride_2015(void) {
+  uint8_t *maxVol    =  (uint8_t*)um_data->u_data[6];
+  uint8_t *binNum    =  (uint8_t*)um_data->u_data[7];
+  float   volumeSmth   = *(float*)  um_data->u_data[0];
+
```
please use only one line to separate parts. More blank lines usually make your code less readable.

tonyxforce committed:
Code improvement changes

tonyxforce commented:
All things you noted above should be fixed in the latest commit

tonyxforce committed:
Merge branch 'MoonModules:mdev' into mdev

tonyxforce committed:
Merge branch 'mdev' of https://github.com/tonyxforce/WLED into mdev

softhack007 commented on wled00/FX.cpp:
```diff
@@ -1915,6 +1915,78 @@ uint16_t mode_pride_2015(void) {
+
+uint16_t mode_jbl() {
+
+  static uint16_t mode_jbl_pos = 0;
```
you could put this into SEGENV.step, to be safe in case of parallel running instances of your effect.


The recommended way to initialize "aux" vars - if you need it - is like this 
```C++
  // this part only runs once, when the effect was started
  if (SEGENV.call == 0) {
    SEGMENT.fill(BLACK);    // clear LEDs
    SEGENV.aux0 = 0;
    SEGENV.aux1 = 0;
    SEGENV.step = 0;
  }
```

tonyxforce replied:
does SEGENV.step change anywhere else?

softhack007 commented:
> All things you noted above should be fixed in the latest commit

thanks, looks much cleaner already :+1

tonyxforce committed:
Minor code modifications/clean up

tonyxforce committed:
Merge branch 'MoonModules:mdev' into mdev

tonyxforce commented:
After uploading the "new" code on my esp32, it looks like the animation is much slower than before (with the [initial](https://github.com/MoonModules/WLED/pull/67/commits/d64b6b3511de98daaf472de9409d81e4306f4f2c) commit)

tonyxforce committed:
Fixing effect speed issues

softhack007 commented:
If you use latest mdev, it could be due to a bug in audioreactive, that caused the microphone to switch of and you would only see "simulated sound".

Please try again with the latest source code. If your effect still seems too slow, please verify that you compare at similar FPS (led settings "target fps"). I can take a closer look on the weekend, maybe there is some timing problem due to integer arithmetics.

softhack007 commented on .vscode/extensions.json:
This seems like a commit accident, please restore the original file.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -1941,7 +2019,7 @@ static const char _data_FX_MODE_JUGGLE[] PROGMEM = "Juggle@!,Trail;;!;;sx=64,ix=
 
 uint16_t mode_palette() {
-  uint16_t counter = 0;
+      uint16_t counter = 0;
```
Same here, this seems like not related to your new effect.

softhack007 commented:
As a minor thing, please think about the name of your effect 
... JBL is a speaker manufacturer, and I don't want to get into conflict with them due to abuse of brand names ...
How about calling it "partybox" or something else, best would be if the name gives people and idea already about what they might get.

softhack007 commented:
"Party Beatbox" maybe?

tonyxforce commented:
how exactly should i refer to [this](https://youtu.be/hnvPLbgRrRE?t=32) effect? I plan to make the other effects this speaker has

tonyxforce committed:
Fix deleted file

tonyxforce committed:
Renamed the effect, biased the sliders, so it
looks best with all sliders in the middle,
added option to change the effect speed multiplier

tonyxforce committed:
Bias value changes

tonyxforce committed:
Merge branch 'mdev' of https://github.com/tonyxforce/WLED into mdev

softhack007 commented:
> how exactly should i refer to [this](https://youtu.be/hnvPLbgRrRE?t=32) effect? I plan to make the other effects this speaker has

Well one of the effects looks exactly like out "gravimeter" (mirrored) just without the peak dot. So you don't need to replicate that one.

Not sure about the others - these rotating rings looks ugly to me, maybe call that "party jerk" :-)

In fact it's your call to find nice names, I'm just saying avoid the word JBL to keep JBL legal far away from us.

tonyxforce commented:
> I'm just saying avoid the word JBL to keep JBL legal far away from us.
Yeah i was thinking about that too

> maybe call that "party jerk"
Party jerk it is

> I think it's almost the hardest part about it

tonyxforce committed:
Fixed possible naming legal issues

tonyxforce committed:
Merge branch 'MoonModules:mdev' into mdev

tonyxforce commented:
are there any more problems or potential enchancements with this effect?

tonyxforce committed:
Merge branch 'MoonModules:mdev' into mdev

softhack007 commented:
> are there any more problems or potential enchancements with this effect?

Give me a week or so to come back to your effect. I was very busy the last weeks with helping to push out upstream -b5.

tonyxforce commented:
> > are there any more problems or potential enchancements with this effect?
> 
> Give me a week or so to come back to your effect. I was very busy the last weeks with helping to push out upstream -b5.
> 

Okay, no problem take your time, i was just wondering

softhack007 commented:
Hi, I'm very busy@job currently, will come back to this PR in October.

tonyxforce committed:
Merge branch 'MoonModules:mdev' into mdev

softhack007 commented:
Hi, I'll merge this PR. I have a few ideas for improvement, but that can also be done later:
* Option to use Sound "Pressure" instead of "Volume"
* make the effect time-dependant, instead of being frame-rate dependant
* and some minor optimisations

this way your new effect will make it into the next MM release.
